### ISSUE TITLE
IMF data request error

### ISSUE THREAD

**Issue by dlevanoni** [dr-leo/pandaSDMX#119]
I'm trying to get pandasdmx (v0.9, using pandas 0.23 and python 3.7) up and running for some IMF data requests. In particular, accessing the BOP (BPM6) series.

```
#!/usr/local/bin/python
import pandas as pd
import pandasdmx

countries = {'Australia': 'AU', 'Austria': 'OE', 'Belgium': 'BE', 'Canada': 'CA', 'Hong Kong': 'HK', 'Denmark': 'DK', 'Euro area': 'EM', 'Finland': 'FN', 'France': 'FR', 'Germany': 'BD', 'Greece': 'GR
', 'Ireland': 'IR', 'Israel': 'IS', 'Italy': 'IT', 'Japan': 'JP', 'Netherlands': 'NL', 'New Zealand': 'NZ', 'Norway': 'NW', 'Portugal': 'PT', 'Singapore': 'SG', 'Spain': 'ES', 'Sweden': 'SE', 'Switzer
land': 'CH', 'United Kingdom': 'UK', 'United States': 'US'}

imf = pandasdmx.Request('IMF_SDMXCENTRAL')
bop_flow = imf.dataflow('BOP_BPM6')
dsd = bop_flow.dataflow.BOP_BPM6.structure()

# Get a list of regions/areas from country list
areas = bop_flow.write().codelist.loc['CL_REF_AREA']
areas = areas[areas['name'].isin(countries.keys())]

# Get  a list of indicators for each of the specific BOP series of interest
codes = bop_flow.write().codelist.loc['CL_INDICATOR']
codes = codes[codes['name'].str.contains('Balance of Payments', case=False) & codes['name'].str.contains('Financial Account', case=False) & codes['name'].str.contains('Portfolio Investment', case=False) & codes['name'].str.contains('Net acquisition of financial assets', case=False) & codes['name'].str.contains('US Dollars', case=False)]
codes = codes[codes.index.str.contains('BP6')]

# Pull down data into a dataframe
for code in list(codes.index):
resp = imf.data(resource_id='BOP_BPM6', key={'INDICATOR': code, 'REF_AREA': list(areas)}, dsd=dsd)
data = resp.data
quarterly = (s for s in data.series if s.key.FREQ == 'D')
df = resp.write(quarterly)
```

The error I get is:

```
requests.exceptions.HTTPError: 404 Client Error:  for url: https://sdmxcentral.imf.org/ws/public/sdmxapi/rest/data/BOP_BPM6?detail=serieskeysonly


```
from this line:

```
Traceback (most recent call last):
  File "./IMFdata.py", line 59, in <module>
    resp = imf.data(resource_id='BOP_BPM6', key={'INDICATOR': code, 'REF_AREA': list(areas)}, dsd=dsd)
```

**Comment by khaeru** [comment]
@dlevanoni—thanks for the bug report. I modified your code to run against pandaSDMX 1.0.0b2, which was just released:
```python
import pandasdmx as sdmx


countries = {
    'Australia': 'AU',
    'Austria': 'OE',
    'Belgium': 'BE',
    'Canada': 'CA',
    'Hong Kong': 'HK',
    'Denmark': 'DK',
    'Euro area': 'EM',
    'Finland': 'FN',
    'France': 'FR',
    'Germany': 'BD',
    'Greece': 'GR',
    'Ireland': 'IR',
    'Israel': 'IS',
    'Italy': 'IT',
    'Japan': 'JP',
    'Netherlands': 'NL',
    'New Zealand': 'NZ',
    'Norway': 'NW',
    'Portugal': 'PT',
    'Singapore': 'SG',
    'Spain': 'ES',
    'Sweden': 'SE',
    'Switzerland': 'CH',
    'United Kingdom': 'UK',
    'United States': 'US',
}

imf = sdmx.Request('IMF')
bop_flow = imf.dataflow('BOP_BPM6')
dsd = bop_flow.dataflow.BOP_BPM6.structure

# Get a list of regions/areas from country list
areas = sdmx.to_pandas(bop_flow.codelist.CL_REF_AREA)
key = dict(
    REF_AREA=areas[areas.isin(countries.keys())].index.to_list(),
    FREQ='D',
    )

print(key)

# Get  a list of indicators for each of the specific BOP series of interest
indicators = sdmx.to_pandas(bop_flow.codelist.CL_INDICATOR)

# Assemble a mask
mask = indicators.index.str.contains('BP6')

# Intersection of the codes where the name contains each of these strings
for s in ['Balance of Payments', 'Financial Account', 'Portfolio Investment',
          'Net acquisition of financial assets', 'US Dollars']:
    mask &= indicators['name'].str.contains(s, case=False)

indicators = indicators[mask].index.to_list()

print('INDICATOR:', indicators)

# Pull down data into a dataframe
for indicator in indicators:
    key.update({'INDICATOR': [indicator]})
    key = 'BOP6.SG.BFPA_BP6_USD.W1.D.2010'
    data = imf.data('BOP_BPM6', key=key, dsd=dsd).data[0]
    df = sdmx.to_pandas(data)
    print(df)
```

However, I'm finding the IMF is returning SDMX error code 100 ([“no results found”](https://github.com/sdmx-twg/sdmx-rest/blob/master/v2_1/ws/rest/docs/4_7_errors.md#no-results-found---100)) for correctly formatted query strings, even an empty key: `https://sdmxcentral.imf.org/ws/public/sdmxapi/rest/data/BOP_BPM6/....`

I also find that from the IMF's web interface at https://sdmxcentral.imf.org prompts for log in if I click on Data → Data Dissemination → Browse Data, or Web Service → Data. I am not even certain they are providing public access to the data, or how to register. @dr-leo, any thoughts?

**Comment by dr-leo** [comment]
Sorry for the delayed response.

A high-level research suggests that the IMF has made changes to the SDMXCentral API. See the [user guide](http://dsbb.imf.org/content/Pdfs/IMFSDMXCentral%20UserGuide_Published%2010_17_2019.pdf). This guide does not talk about data requests at all. Only structure queries seem to be supported. I read in another guide which may or may not be up to date, that data requests are still possible. But I found it only through google, not the SDMXCentral website itself. That guide, however, talks about user authentication. There may be a chance that data requests may be restricted to registered users now. 

This topic needs further research.

**Comment by KraxelHuber** [comment]
I am also running into an `404 Client Error`.

```
import pandasdmx as sdmx

print(sdmx.__version__)

data_source = "IMF"
src = sdmx.Request(data_source, backend="sqlite", fast_save=True, expire_after=600)

src_msg = src.dataflow("NAG_NSTD")

src_flow = src_msg.dataflow.NAG_NSTD

dsd = src_flow.structure

keys = {"DATA_DOMAIN": "NAG", "INDICATOR": "SUR_NGDPVA_R_ISIC3_TOTAL_G_XDC"}
params = {"StartPeriod": "2000"}

out = src.data("NAG_NSTD", key=keys, params=params, dsd=dsd)

```

Traceback:
```
1.0.0
Traceback (most recent call last):
  File "<input>", line 1, in <module>
  File "C:\Program Files\JetBrains\PyCharm Community Edition 2019.3.3\plugins\python-ce\helpers\pydev\_pydev_bundle\pydev_umd.py", line 197, in runfile
    pydev_imports.execfile(filename, global_vars, local_vars)  # execute the script
  File "C:\Program Files\JetBrains\PyCharm Community Edition 2019.3.3\plugins\python-ce\helpers\pydev\_pydev_imps\_pydev_execfile.py", line 18, in execfile
    exec(compile(contents+"\n", file, 'exec'), glob, loc)
  File "C:/Users/andi/PycharmProjects/sdmx/sdmx.py", line 17, in <module>
    out = src.data("NAG_NSTD", key=keys, params=params, dsd=dsd)
  File "C:\Users\andi\AppData\Local\Programs\Python\Python38\lib\site-packages\pandasdmx\api.py", line 390, in get
    response.raise_for_status()
  File "C:\Users\andi\AppData\Local\Programs\Python\Python38\lib\site-packages\requests\models.py", line 941, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 404 Client Error:  for url: https://sdmxcentral.imf.org/ws/public/sdmxapi/rest/data/NAG_NSTD/NAG..SUR_NGDPVA_R_ISIC3_TOTAL_G_XDC...?StartPeriod=2000

```

In the docs, it is mentioned that IMF `Supports series-key-only and hence dataset-based key validation and construction.` I don't fully understand the consequences. Any hint how the code should look like to take this into account is highly appreciated.

**Comment by dr-leo** [comment]
The source of this issue is not understood yet. It is a high priority
to fix this, now that v1.0.0 is out.

Any help would be appreciated.

**Comment by dr-leo** [comment]
I can confirm that IMF SDMXCentral unfortunately no longer supports data queries. It continues to support queries for structural metadata though, however useful this may be to average data consumers.

I will update the docs and possibly pandasdmx.sources package accordingly to avoid the misleading Impression that IMF data can be queried with pandasdmx.

TODO: 

* consider alternative channels to obtain IMF data.
